tensorflow/core/util/use_cudnn.cc

```diff
@@ -56,12 +56,15 @@ bool CudnnUseFrontend() {
 
 // Whether to enable Cudnn runtime compiled kernels which are able to support
 // more general fusion patterns but might increase the warmup time.
-bool CudnnUseRuntime() {
+// TODO(kaixih@nvidia): we can make it default when Cudnn further improves the
+// runtime compilation overhead.
```

Comment:
Is this the TODO, or is the TODO to improve *TensorFlow* so that it avoids pathological behavior?  Like, I imagine that cudnn can only make runtime compilation so fast.

Reply:
Our cudnn team is working on further improving lowering the compilation overhead. Currently, each engine building needs about 1.3-1.5 seconds. Though this could only affect the autotuning, it would bring bad user experience when users are developing and tuning the model. So, here I make this feature as an opt-in.

Aside: for XLA, I think this might be less an issue, since AFAIK, XLA autotunes the conv by building the engine once and picking the fastest one by sweeping them once. But for native TF, we need to build and run the engines in every autotuning steps and promote/demote the best option dynamically.

Reply:
> XLA autotunes the conv by building the engine once and picking the fastest one by sweeping them once. But for native TF, we need to build and run the engines in every autotuning steps and promote/demote the best option dynamically.

Hm, I am not understanding. I think TF has the same problem (or no problem) as XLA?

In XLA today, every time we see a new set of shapes, it's a new XLA compilation.  We recompile from scratch, and do autotuning again.

In TF today, every time we see a new conv shape, we re-autotune it.  Exactly the same behavior as XLA wrt conv autotuning.

So if we can afford to use the cudnn runtime in XLA, why can't we afford to do it in TF?

Reply:
Sorry for the confusion. What I meant to say is for a known shape, the autotuning algorithms of XLA and TF are different.

For example, in TF, we are doing conv for shape A and the typical autotuning is like:
```
Training Step 1:
(1) Query working engines for conv shape A; For example, we get 10 engines.
(2) Run all these 10 engines and pick the best one; For example, we pick the 2nd one.
(3) Cache the 2nd one in the map (count=1), and use the 2nd one in the actual computation.
Training Step 2:
(1) <Same with the above>
(2) <Same with the above>
(3) Promote the 2nd engine in the map (count=2), and still use the 2nd one in the actual computation.
Training Step 3:
(1) <Same with the above>
(2) <Same with the above> But we pick the 1st engine this time
(3) Demote the 2nd engine in the map (count=1), and keep using the 2nd one in the actual computation.
......This will go on until some threshhold is hit, e.g. the count exceeds some value or the total autotuning step reaches some value.
```
The issue of this autotuning algorithm is the step (1) when using the runtime compilation, since (1) needs to query *and* build/compile all the engines. And we don't cache those engines that are not picked by the (2). That is why I say the autotuning steps will be more expensive when runtime compilation is enabled.

As for the XLA, my understanding is that for conv shape A:
```
During XLA compilation, we do
(1) Query working engines for conv shape A; For example, we get 10 engines.
(2) Run all these 10 engines (once, twice or whatever times) and pick the best one; For example, we pick the 2nd one.
And that is all, so we will simply use the 2nd one and cache it/serialize it.
```
If this understanding is correct, the cudnn only builds the engines once in this scenario. That is why I feel XLA may be able to afford it.

Reply:
> For example, in TF, we are doing conv for shape A and the typical autotuning is like:

I see.   I'm looking at this: https://github.com/tensorflow/tensorflow/blob/ba57ae7f24743e684accef3521485a24c1235186/tensorflow/core/kernels/gpu_utils.h#L112

This tries to avoid re-autotuning if the best algorithm is "a lot better" than the alternatives.  But if it's a close call, we may re-autotune many times, and that will be expensive if we're recompiling every time we re-autotune.

I guess we have to fix that logic, so that we choose a good algorithm with high confidence "the first time"?  Could we make *that* the TODO, instead of "fix cudnn"?

Reply:
Yes, that is actually on my TODO list. In the current autotune algorithm, we can already manually limit the re-autotune times by playing with `TF_AUTOTUNE_THRESHOLD` and `TF_AUTOTUNE_MIN_WARMUP_ITERATIONS`. Or as you mentioned, maybe we can add some logic to let the autotuner to smartly and automatically deal with the engines having close perf.  Besides, I am also thinking about to cache the built engines during autotuning or simply disable autotuning for the runtime-compiled engines.

(Also, in https://github.com/tensorflow/tensorflow/pull/56572#discussion_r937419136, all the `count` should be `score`. The `count` is used to avoid too many re-autotuning times.)